src/serde/de.rs

```diff
@@ -0,0 +1,931 @@
+pub fn from_ion<'a, T, S>(s: S) -> IonResult<T>
+where
+    T: Deserialize<'a>,
+    S: IonDataSource,
+{
+    let mut deserializer = Deserializer {
+        reader: ReaderBuilder::new().build(s)?,
+    };
+
+    if StreamItem::Nothing == deserializer.reader.current() {
+        // We're not on a value. Advance the reader.
+        deserializer.reader.next()?;
+    }
+
+    if StreamItem::Nothing == deserializer.reader.current() {
+        // Advancing the reader did nothing; this input contains no values.
+        return IonResult::decoding_error(
+            "The input for deserialization doesn't contain any values",
+        );
+    }
+
+    let result = T::deserialize(&mut deserializer)?;
+    Ok(result)
+}
+
+/// The deserializer for Ion, it doesn't care about
+/// whether the reader is reading binary or text representation.
+#[derive(Debug)]
+pub struct Deserializer<R> {
+    pub(crate) reader: R,
+}
+
+impl<'de, 'a, R> de::Deserializer<'de> for &'a mut Deserializer<R>
+where
+    R: IonReader<Symbol = Symbol, Item = StreamItem>,
+{
+    type Error = IonError;
+
+    fn deserialize_any<V>(self, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: Visitor<'de>,
+    {
+        if let Some(ion_type) = self.reader.ion_type() {
+            match ion_type {
+                IonType::Null => self.deserialize_unit(visitor),
+                IonType::Blob | IonType::Clob => self.deserialize_bytes(visitor),
+                IonType::String | IonType::Symbol => self.deserialize_string(visitor),
+                IonType::Float => self.deserialize_f64(visitor),
+                IonType::Int => self.deserialize_i64(visitor),
+                IonType::Decimal => {
+                    self.deserialize_newtype_struct(TUNNELED_DECIMAL_TYPE_NAME, visitor)
+                }
+                IonType::Bool => self.deserialize_bool(visitor),
+                IonType::List => self.deserialize_seq(visitor),
+                IonType::Struct => self.deserialize_struct("", &[], visitor),
+                IonType::Timestamp => {
+                    self.deserialize_newtype_struct(TUNNELED_TIMESTAMP_TYPE_NAME, visitor)
+                }
+                _ => IonResult::decoding_error("unexpected ion type"),
+            }
+        } else {
+            IonResult::decoding_error("unexpected end of file")
+        }
+    }
+
+    fn deserialize_bool<V>(self, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: Visitor<'de>,
+    {
+        let result = visitor.visit_bool(self.reader.read_bool()?);
+        self.reader.next()?;
+        result
+    }
+
+    fn deserialize_i8<V>(self, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: Visitor<'de>,
+    {
+        let result = visitor.visit_i8(
+            self.reader
+                .read_i64()
+                .map(i8::try_from)?
+                .map_err(|e| IonError::decoding_error(e.to_string()))?,
```

Comment:
For each of these, the `e.to_string()` makes something not super informative for the user: `Err(TryFromIntError(()))`. ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c040d5c87ffc358953831e6830261eca))

I suggest adding your own text:
```
.map_err(|_| IonError::decoding_error(found an integer was out of bounds for an `i8`"))
```